File: shared/src/main/java/de/dreier/mytargets/shared/utils/FileUtils.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.utils;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.content.FileProvider;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.FileChannel;

public class FileUtils {
    public static void copy(File src, File dst) throws IOException {
        FileInputStream inStream = new FileInputStream(src);
        FileOutputStream outStream = new FileOutputStream(dst);
        FileChannel inChannel = inStream.getChannel();
        FileChannel outChannel = outStream.getChannel();
        inChannel.transferTo(0, inChannel.size(), outChannel);
        inStream.close();
        outStream.close();
    }

    public static void copy(InputStream in, File dst) throws IOException {
        copy(in, new FileOutputStream(dst));
    }

    public static void copy(InputStream in, OutputStream out) throws IOException {
        byte[] buf = new byte[1024];
        int len;
        while ((len = in.read(buf)) > 0) {
            out.write(buf, 0, len);
        }
        in.close();
        out.close();
    }

    public static Uri getUriForFile(Context context, File file) {
        String packageName = context.getPackageName();
        String authority = packageName + ".fileprovider";
        return FileProvider.getUriForFile(context, authority, file);
    }

    public static void move(@NonNull File from, @NonNull File to) throws IOException {
        File directory = to.getParentFile();
        if (!directory.exists()) {
            directory.mkdir();
        }
        copy(from, to);
        from.delete();
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/utils/FileUtils.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.utils

import android.content.Context
import android.net.Uri
import android.support.v4.content.FileProvider
import java.io.File
import java.io.IOException

@Throws(IOException::class)
fun File.moveTo(to: File) {
    val directory = to.parentFile
    if (!directory.exists()) {
        directory.mkdir()
    }
    copyTo(to, overwrite = true)
    delete()
}

fun File.toUri(context: Context): Uri {
    val packageName = context.packageName
    val authority = packageName + ".fileprovider"
    return FileProvider.getUriForFile(context, authority, this)
}
